common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java

```diff
@@ -94,6 +95,9 @@
   static final String STOP_ON_FAILURE_KEY = "spark.yarn.shuffle.stopOnFailure";
   private static final boolean DEFAULT_STOP_ON_FAILURE = false;
 
+  // Used by shuffle merge manager to create merged shuffle files.
+  protected static final String APP_BASE_RELATIVE_PATH = "usercache/%s/appcache/%s/";
```

Comment:
This is only used inside `RemoteBlockPushResolver`. Could we declare it as a static global valuable there?

Reply:
Wait, is it the Yarn-specific local dir format? If so, please ignore my above comment.

Reply:
Yes it a Yarn-specific format for local dirs.

Reply:
yeah, it does feel a bit odd here to hard code this here.   the code inside RemoteBlockPushResolver assumes it has a specific format of with the user and appid as well. I don't have any other great solution to get this from yarn, but I think it should be documented what the %s values are and assumed to be when passed into the class.

Reply:
While making changes to support sub-dir, I realized that `RemoteBlockPushResolver` doesn't really need this 
`usercache/%s/appcache/%s` format.
In one of our older versions when we were **not** using the  `executorShuffleInfo` from the `registerExecutor` message, we added this. However, now it uses the `localDirs` from `executorShuffleInfo` which are paths of the `blockManagerDirs`. The current implementation finds the root local dir by `localDir.substring(0, "usercache/{userId}/appcache/{appId/")` which seems un-necessary. What we need is to find the parent directory of each `blockManagerDir` and construct the path of each `merge_manager` directory, which is a sibling of `blockManagerDir`. Even though this is not that clean but it is similar to what we have now and doesn't change the `registerExecutor` protocol.

Although this may increase the memory footprint of `appPathsInfo` map slightly.
In an `appPathInfo`, instead of saving the complete local paths, for example 
`[localDir1/usercache/{userId}/appcache/{appId}/merge_manager, localDir2/usercache/{userId}/appcache/{appId}/merge_manager/]`, currently we only save
`[localDir1, localDir2]`.  While constructing the target file path we add the sub-part `usercache/{userId}/appcache/{appId}`.
The target file path would be constructing using these parts: `{localDir}/{usercache/%s/appcache/%s/merge_manager}/filename`. 
I do think this increase will not be that significant. 
It would be much better to remove this assumption on the path format.

@Victsm @zhouyejoe  @tgravescs @Ngone51 @attilapiros @mridulm

Reply:
This goes back to the conversation we had in #29855 (https://github.com/apache/spark/pull/29855/files#r497077157)

The assumption right now is that we can use appId and user Id to properly distinguish the local dirs given different cluster schedulers.
In the case of YARN, this information is provided through the following API during application registration:
https://github.com/apache/spark/blob/f284218dae23bf91e72e221943188cdb85e13dac/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/MergedShuffleFileManager.java#L58-L66

During executor registration, the `blockManagerDir` local dir paths information is further provided to `RemoteBlockPushResolver` through the following API:
https://github.com/apache/spark/blob/f284218dae23bf91e72e221943188cdb85e13dac/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/MergedShuffleFileManager.java#L68-L77
Since block merge dirs are per-application instead of per-executor, we only pick the local dir information for the first executor of a given application registered with the shuffle service.

Having the common dir path pattern provided by `YarnShuffleService` to `RemoteBlockPushResolver` would then allow `RemoteBlockPushResolver` to construct the local dirs used for storing merged shuffle files from the `blockManagerDir` local dirs provided during executor registration.

As @otterc mentioned, we don't really need the common dir path pattern if we just store the full local dir paths in the hashmap.
This would also simplify the application registration API in `MergeShuffleFileManager` to remove the user ID field, which could make it more applicable to other cluster schedulers.

Reply:
Also I think we don't need a separate `registerApplication` API anymore. It's just creating an empty `appsPathInfo` for the appId and adds it to the `appsPathInfo`. This could be done in `registerExecutor`. Even with `ExternalShuffleBlockResolver` there isn't any registerApplication API. 
My next commit will remove this as well. 
cc. @Victsm